src/Exception/DriverException.php

```diff
@@ -32,4 +47,43 @@ public function getSQLState(): ?string
 
         return $previous->getSQLState();
     }
+
+    /**
+     * @internal This is only meant to be called by the DBAL.
+     *
+     * @param array<mixed>                $params
+     * @param array<Type|int|string|null> $types
+     *
+     * @psalm-suppress ImpurePropertyAssignment
+     */
+    public function withQueryParams(array $params, array $types): self
```

Comment:
I don't like it either, that's just the solution that required the least amount of modifications. I'll try to rework this according to your suggestions!

Reply:
Actually we can't just introduce an `Exception\Query` class: all subclasses of `DBAL\Exception\DriverException` (created in the exception converters) may potentially need the SQL, params & types.

**I think adding the properties to `DriverException` is therefore the way to go**, but I will need to change the `ExceptionConverter::convert()` signature to accept the SQL, params & types as well, which is cumbersome. Is that acceptable?

An alternative to avoid the bloat while still passing all the information is to replace `string $message` in both `ExceptionConverter::convert()` and the `DriverException` constructor with an object that will contain the exception message, the SQL, the params & the types:

```php
class DriverExceptionInfo {
    private string $message;
    private string $sql;
    private array $params;
    private array $types;
}
```

Another alternative would be to return a class-string in `ExceptionConverter::convert()` instead of a class instance (as I don't think exception converters do anything else than `new *Exception()`, do they?), this way we don't need to pass the SQL params etc. to the converters at all.

Thoughts?

Reply:
> ```php
> class DriverExceptionInfo {
>     private string $message;
>     private string $sql;
>     private array $params;
>     private array $types;
> }
> ```

This is what I meant by the `Query` class:
```php
class Query {
    private string $sql;
    private array $params;
    private array $types;
}
```

Reply:
Sorry, I read your comment too fast. Do you want to follow this route then? (this will be a breaking change to the `ExceptionConverter`)

Reply:
Yeah, the driver converter will have to accept an optional query. I don't think it's a problem.

Reply:
Done!